src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,143 @@
+-- | This module contains a compiler pass that analyses the right hand sides of
+--   function declaration and introduces @let@ expression with new variables
+--   for each variable that occurs more than once on the right hand sides.
+--
+--   = Examples
+--
+--   == Example 1
+--
+--   > twice :: Integer -> Integer
+--   > twice ( x :: Integer ) = x + x
+--
+--   Should be transformed into
+--
+--   > twice :: Integer -> Maybe Integer
+--   > twice ( x :: Integer ) = let ( y :: Integer ) = x in y + y
+--
+--   == Example 2
+--
+--   data Maybe a = Just a | Nothing
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe ( mx :: Maybe Integer ) = case ( mx :: Maybe Integer ) of
+--   >   { Nothing -> Nothing
+--   >   ; Just x -> Just (x + x)
+--   >   }
+--
+--   Should be transformed into
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe ( mx :: Maybe Integer ) = case ( mx :: Maybe Integer ) of
+--   >   { Nothing -> Nothing
+--   >   ; Just x -> let y = x in Just ( y + y )
+--   >   }
```

Comment:
The following formatting is more consistent with the documentation of the other passes. I don't think that the definition of the `Maybe` data type needs to be recapitulated.

```suggestion
--   = Examples
--
--   == Example 1
--
--   > twice :: Integer -> Integer
--   > twice (x :: Integer) = x + x
--
--   Should be transformed into
--
--   > twice :: Integer -> Maybe Integer
--   > twice (x :: Integer) = let (y :: Integer) = x in y + y
--
--   == Example 2
--
--   > twiceMaybe :: Maybe Integer -> Maybe Integer
--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
--   >     Nothing -> Nothing;
--   >     Just x  -> Just (x + x)
--   >   }
--
--   Should be transformed into
--
--   > twiceMaybe :: Maybe Integer -> Maybe Integer
--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
--   >     Nothing -> Nothing;
--   >     Just x  -> let y = x in Just (y + y)
--   >   }
```